rfd/0184-agent-auto-updates.md

```diff
@@ -0,0 +1,1640 @@
+   │  └── etc
+   │     └── systemd
+   │        └── teleport.service
+   ├── system # if installed via OS package
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-update
+   │  │  └── teleport
+   │  └── etc
+   │     └── systemd
+   │        └── teleport.service
+   └── update.yaml
+$ ls -l /usr/local/bin/tsh
+/usr/local/bin/tsh -> /var/lib/teleport/versions/15.0.0/bin/tsh
+$ ls -l /usr/local/bin/tbot
+/usr/local/bin/tbot -> /var/lib/teleport/versions/15.0.0/bin/tbot
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport-update
+/usr/local/bin/teleport-update -> /var/lib/teleport/versions/15.0.0/bin/teleport-update
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+##### update.yaml
+
+This file stores configuration for `teleport-update`.
+
+All updates are applied atomically using renameio.
+
+```
+version: v1
+kind: update_config
+spec:
+  # proxy specifies the Teleport proxy address to retrieve the agent version and update configuration from.
+  proxy: mytenant.teleport.sh
+  # group specifies the update group
+  group: staging
+  # url_template specifies a custom URL template for downloading Teleport.
+  # url_template: ""
+  # enabled specifies whether auto-updates are enabled, i.e., whether teleport-update update is allowed to update the agent.
+  enabled: true
+status:
+  # start_time specifies the start time of the most recent update.
+  start_time: 2020-12-09T16:09:53+00:00
+  # active_version specifies the active (symlinked) deployment of the teleport agent.
+  active_version: 15.1.1
+  # version_history specifies the previous deployed versions, in order by recency.
+  version_history: ["15.1.3", "15.0.4"]
+  # rollback specifies whether the most recent version was deployed by an automated rollback.
+  rollback: true
+  # error specifies the last error encounted
+  error: "" 
+```
+
+##### backup.yaml
+
+This file stores metadata about an individual backup of the Teleport agent's sqlite DB.
+
+```
+version: v1
+kind: db_backup
+spec:
+  # proxy address from the backup
+  proxy: mytenant.teleport.sh
+  # version from the backup
+  version: 15.1.0
+  # time the backup was created
+  creation_time: 2020-12-09T16:09:53+00:00
+```
+
+#### Runtime
+
+The `teleport-update` binary will run as a periodically executing systemd service which runs every 10 minutes.
+The systemd service will run:
+```shell
+$ teleport-update update
+```
+
+After it is installed, the `update` subcommand will no-op when executed until configured with the `teleport-update` command:
+```shell
+$ teleport-update enable --proxy mytenant.teleport.sh --group staging
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used, if present.
```

Comment:
It will have the same interface and default for specifying the data dir as the teleport binary:
https://github.com/gravitational/teleport/pull/46418/files#diff-d099d69928110c77b90c52fa55c9d973aa6070b7f1653c3eb6f6cd3c65352643R124